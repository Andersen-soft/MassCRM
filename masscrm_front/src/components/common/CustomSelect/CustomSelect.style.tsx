import { createStyles, makeStyles } from '@material-ui/core';

export const selectStyle = makeStyles(() =>
  createStyles({
    multiCount: {
      width: '24px',
      height: '24px',
      background: '#F4F6FA',
      boxShadow: '0px 1px 1px rgba(0, 25, 91, 0.15)',
      borderRadius: '4px',
      textAlign: 'center',
      padding: '5px 1px',
      fontSize: '0.85em',
      boxSizing: 'border-box',
      position: 'absolute',
      top: '17%',
      right: '26%',
      color: '#939BB2'
    },
    customSelectIcon: {
      color: '#69738F'
    },
    customSelectOutlined: {
      paddingRight: 0,
      width: '85%',
      background: '#fff !important',
      textAlign: 'left'
    },
    Paper: {
      marginTop: '50px',
      '& .MuiMenu-paper': {
        maxHeight: '200px'
      }
    },
    ListItemBtn: {
      '&:hover': {
        background: '#F9F9FA'
      }
    },
    InputAdornmentPositionEnd: {
      marginLeft: '-65px'
    },
    IconBtnEdgeEnd: {
      left: '15px',
      padding: '0',
      marginRight: '0',
      '&:hover': {
        background: 'none'
      }
    },
    InputLabelOutlined: {
      transform: 'translate(14px, 14px) scale(1)',
      color: '#939BB2',
      fontSize: '1em'
    },
    InputLabelShrink: {
      background: '#fff'
    },
    FormLabel: {
      width: '230px',
      minHeight: '40px',
      boxSizing: 'border-box',
      '& .MuiOutlinedInput-root': {
        height: '100%'
      },
      '& .MuiFormLabel-root.Mui-focused': {
        color: '#939BB2',
        transform: 'translate(14px, -4px) scale(1)',
        fontSize: '0.85em'
      },
      '& .MuiOutlinedInput-notchedOutline': {
        borderColor: '#E1E5ED'
      },
      '.MuiOutlinedInput-root:hover .MuiOutlinedInput-notchedOutline, & .MuiOutlinedInput-root.Mui-focused .MuiOutlinedInput-notchedOutline': {
        borderColor: '#B6BECF',
        borderWidth: '1px'
      },
      '& .MuiSelect-nativeInput': {
        color: '#212121',
        top: 0,
        opacity: 1,
        border: 'none',
        width: '50%',
        left: '10px'
      },
      '& .MuiInputAdornment-root': {
        marginLeft: '-42px'
      }
    },
    required: {
      '& .MuiInputLabel-outlined::before': {
        content: '"*"'
      }
    },
    error: {
      '& .MuiOutlinedInput-root, & .MuiOutlinedInput-notchedOutline': {
        borderColor: '#FD5757'
      }
    },
    SvgIconFontSizeSmall: {
      fontSize: '1rem'
    },
    MultiIconButtonLabel: {
      background: '#fff',
      zIndex: 2,
      '&&:hover': {
        background: '#fff'
      }
    },
    IconButtonRoot: {
      padding: '0 ',
      marginRight: '0'
    },
    wrap: {
      position: 'relative'
    }
  })
);
