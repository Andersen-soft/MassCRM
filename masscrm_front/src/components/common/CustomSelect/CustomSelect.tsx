import React, { ChangeEvent, FC, useMemo } from 'react';
import {
  FormControl,
  IconButton,
  InputAdornment,
  MenuItem,
  Select,
  InputLabel,
  Box
} from '@material-ui/core';
import 'src/styles/main.scss';
import { Close } from '@material-ui/icons';
import { inputStyle } from 'src/styles/CommonInput.style';
import { selectStyle } from './CustomSelect.style';
import { ICustomSelectProps } from './interfaces';

export const CustomSelect: FC<ICustomSelectProps> = ({
  multi,
  placeholder,
  value = undefined,
  items,
  required,
  errorMessage,
  name,
  width,
  disabled,
  hideClearBtn,
  eventInResult,
  onChange,
  inputClassName,
  information
}) => {
  const inputStyles = inputStyle();
  const styles = selectStyle();
  const cleanField = () => {
    onChange();
  };

  const onChangeHandler = (
    event: ChangeEvent<{ name?: string | undefined; value: unknown }>
  ) => {
    onChange(eventInResult ? event : (event.target?.value as Array<string>));
  };

  const classInput = useMemo(
    () =>
      `${inputStyles.input} ${styles.FormLabel} ${required &&
        styles.required} ${errorMessage && styles.error} ${inputClassName &&
        inputClassName}`,
    [required, errorMessage, inputClassName]
  );

  const classLabel = useMemo(
    () => ({
      outlined: styles.InputLabelOutlined,
      shrink: styles.InputLabelShrink
    }),
    []
  );

  const classSelect = useMemo(
    () => ({
      icon: styles.customSelectIcon,
      outlined: styles.customSelectOutlined
    }),
    []
  );

  const classIcon = useMemo(
    () => ({
      edgeEnd: styles.IconBtnEdgeEnd,
      label: multi ? styles.MultiIconButtonLabel : '',
      root: styles.IconButtonRoot
    }),
    [multi]
  );

  const classClose = useMemo(
    () => ({ fontSizeSmall: styles.SvgIconFontSizeSmall }),
    []
  );

  return (
    <div className={styles.wrap}>
      <FormControl
        style={{ width: width || '230px' }}
        variant='outlined'
        classes={{
          root: classInput
        }}
      >
        <InputLabel id={name} classes={classLabel}>
          {placeholder}
        </InputLabel>
        <Select
          classes={classSelect}
          MenuProps={{
            className: styles.Paper
          }}
          disabled={disabled}
          inputProps={{ id: name, name, className: 'selectInput' }}
          onChange={onChangeHandler}
          value={value || []}
          required={required}
          renderValue={
            multi
              ? selected => (
                  <div className={styles.multiCount}>
                    {(selected as Array<string>).length}
                  </div>
                )
              : undefined
          }
          multiple={multi}
          startAdornment={information}
          endAdornment={
            Boolean(value) && !hideClearBtn ? (
              <InputAdornment
                position='end'
                classes={{
                  positionEnd: styles.InputAdornmentPositionEnd
                }}
              >
                <IconButton
                  aria-label=''
                  onClick={cleanField}
                  onMouseDown={() => false}
                  edge='end'
                  classes={classIcon}
                >
                  <Close fontSize='small' classes={classClose} />
                </IconButton>
              </InputAdornment>
            ) : (
              undefined
            )
          }
        >
          {items.map(item => (
            <MenuItem value={item} key={item} className={styles.ListItemBtn}>
              {item}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
      {errorMessage && <Box className={inputStyles.error}>{errorMessage}</Box>}
    </div>
  );
};
