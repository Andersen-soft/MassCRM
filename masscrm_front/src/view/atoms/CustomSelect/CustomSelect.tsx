import React, { ChangeEvent, FC, ReactNode, useMemo } from 'react';
import {
  FormControl,
  IconButton,
  InputAdornment,
  MenuItem,
  Select,
  InputLabel,
  Box
} from '@material-ui/core';
import 'src/styles/main.scss';
import { Close } from '@material-ui/icons';
import { commonInputStyles } from 'src/styles';
import { BOTTOM, CENTER, getPositionConfig, TOP } from 'src/constants';
import { useStyles } from './CustomSelect.styles';

interface IProps {
  items: string[];
  placeholder: string;
  multi?: boolean;
  value?: string | string[];
  name?: string;
  required?: boolean;
  errorMessage?: string | string[];
  width?: string;
  disabled?: boolean;
  hideClearBtn?: boolean;
  eventInResult?: boolean;
  onChange: Function;
  inputClassName?: string;
  information?: ReactNode;
}

export const CustomSelect: FC<IProps> = ({
  multi,
  placeholder,
  value = [],
  items,
  required,
  errorMessage,
  name,
  width = '230px',
  disabled,
  hideClearBtn,
  eventInResult,
  onChange,
  inputClassName,
  information
}) => {
  const commonInputClasses = commonInputStyles();
  const styles = useStyles();

  const cleanField = () => {
    onChange();
  };

  const onChangeHandler = (
    event: ChangeEvent<{ name?: string; value: unknown }>
  ) => {
    onChange(eventInResult ? event : (event.target.value as string[]));
  };

  const classInput = useMemo(
    () =>
      `${commonInputClasses.input} ${styles.FormLabel} ${required &&
        styles.required} ${errorMessage && styles.error} ${inputClassName &&
        inputClassName}`,
    [required, errorMessage, inputClassName]
  );

  const classLabel = useMemo(
    () => ({
      outlined: styles.InputLabelOutlined,
      shrink: styles.InputLabelShrink
    }),
    []
  );

  const classSelect = useMemo(
    () => ({
      icon: styles.customSelectIcon,
      outlined: styles.customSelectOutlined
    }),
    []
  );

  const classIcon = useMemo(
    () => ({
      edgeEnd: styles.IconBtnEdgeEnd,
      label: multi ? styles.MultiIconButtonLabel : '',
      root: styles.IconButtonRoot
    }),
    [multi]
  );

  const classClose = useMemo(
    () => ({ fontSizeSmall: styles.SvgIconFontSizeSmall }),
    []
  );

  return (
    <div className={styles.wrap}>
      <FormControl
        style={{ width }}
        variant='outlined'
        classes={{
          root: classInput
        }}
      >
        <InputLabel id={name} classes={classLabel}>
          {placeholder}
        </InputLabel>
        <Select
          classes={classSelect}
          MenuProps={{
            className: styles.Paper,
            getContentAnchorEl: null,
            anchorOrigin: {
              ...getPositionConfig(BOTTOM, CENTER)
            },
            transformOrigin: {
              ...getPositionConfig(TOP, CENTER)
            }
          }}
          disabled={disabled}
          inputProps={{ id: name, name, className: 'selectInput' }}
          onChange={onChangeHandler}
          value={value}
          required={required}
          renderValue={
            multi
              ? selected => (
                  <div className={styles.multiCount}>
                    {(selected as string[]).length}
                  </div>
                )
              : undefined
          }
          multiple={multi}
          startAdornment={information}
          endAdornment={
            value && !hideClearBtn ? (
              <InputAdornment
                position='end'
                classes={{
                  positionEnd: styles.InputAdornmentPositionEnd
                }}
              >
                <IconButton
                  aria-label=''
                  onClick={cleanField}
                  onMouseDown={() => false}
                  edge='end'
                  classes={classIcon}
                >
                  <Close fontSize='small' classes={classClose} />
                </IconButton>
              </InputAdornment>
            ) : (
              undefined
            )
          }
        >
          {items.map(item => (
            <MenuItem value={item} key={item} className={styles.ListItemBtn}>
              {item}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
      {errorMessage && (
        <Box className={commonInputClasses.error}>{errorMessage}</Box>
      )}
    </div>
  );
};
