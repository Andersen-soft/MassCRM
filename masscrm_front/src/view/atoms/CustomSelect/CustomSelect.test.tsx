import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { CustomSelect } from './CustomSelect';

describe('CustomSelect', () => {
  const mockCallBack = jest.fn();

  const customSelect = shallow(
    <CustomSelect
      onChange={mockCallBack}
      items={['A', 'B', 'C', 'D', 'E']}
      placeholder='test'
      multi
    />
  );

  it('MultiSelectFilter should render correctly', () => {
    expect(shallowToJson(customSelect)).toMatchSnapshot();
  });
});
