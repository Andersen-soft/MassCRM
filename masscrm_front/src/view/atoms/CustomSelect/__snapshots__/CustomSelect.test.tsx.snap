// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSelect MultiSelectFilter should render correctly 1`] = `
<div
  className="makeStyles-wrap-27"
>
  <WithStyles(ForwardRef(FormControl))
    classes={
      Object {
        "root": "makeStyles-input-1 makeStyles-FormLabel-21 undefined undefined undefined",
      }
    }
    style={
      Object {
        "width": "230px",
      }
    }
    variant="outlined"
  >
    <WithStyles(ForwardRef(InputLabel))
      classes={
        Object {
          "outlined": "makeStyles-InputLabelOutlined-19",
          "shrink": "makeStyles-InputLabelShrink-20",
        }
      }
    >
      test
    </WithStyles(ForwardRef(InputLabel))>
    <WithStyles(ForwardRef(Select))
      MenuProps={
        Object {
          "anchorOrigin": Object {
            "horizontal": "center",
            "vertical": "bottom",
          },
          "className": "makeStyles-Paper-15",
          "getContentAnchorEl": null,
          "transformOrigin": Object {
            "horizontal": "center",
            "vertical": "top",
          },
        }
      }
      classes={
        Object {
          "icon": "makeStyles-customSelectIcon-13",
          "outlined": "makeStyles-customSelectOutlined-14",
        }
      }
      endAdornment={
        <ForwardRef(WithStyles)
          classes={
            Object {
              "positionEnd": "makeStyles-InputAdornmentPositionEnd-17",
            }
          }
          position="end"
        >
          <ForwardRef(WithStyles)
            aria-label=""
            classes={
              Object {
                "edgeEnd": "makeStyles-IconBtnEdgeEnd-18",
                "label": "makeStyles-MultiIconButtonLabel-25",
                "root": "makeStyles-IconButtonRoot-26",
              }
            }
            edge="end"
            onClick={[Function]}
            onMouseDown={[Function]}
          >
            <UNDEFINED
              classes={
                Object {
                  "fontSizeSmall": "makeStyles-SvgIconFontSizeSmall-24",
                }
              }
              fontSize="small"
            />
          </ForwardRef(WithStyles)>
        </ForwardRef(WithStyles)>
      }
      inputProps={
        Object {
          "className": "selectInput",
          "id": undefined,
          "name": undefined,
        }
      }
      multiple={true}
      onChange={[Function]}
      renderValue={[Function]}
      value={Array []}
    >
      <WithStyles(ForwardRef(MenuItem))
        className="makeStyles-ListItemBtn-16"
        key="A"
        value="A"
      >
        A
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem))
        className="makeStyles-ListItemBtn-16"
        key="B"
        value="B"
      >
        B
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem))
        className="makeStyles-ListItemBtn-16"
        key="C"
        value="C"
      >
        C
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem))
        className="makeStyles-ListItemBtn-16"
        key="D"
        value="D"
      >
        D
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem))
        className="makeStyles-ListItemBtn-16"
        key="E"
        value="E"
      >
        E
      </WithStyles(ForwardRef(MenuItem))>
    </WithStyles(ForwardRef(Select))>
  </WithStyles(ForwardRef(FormControl))>
</div>
`;
